### PR TITLE
Optimizing ChunkingCookieManager

### PR DESCRIPTION
**Pull request by uabarahona**
# Summary 
These changes are to improve the performance for the operation done by ChunkingCookieManager

#### AppendCookies (~2x win)

|        Method |     Mean |     Error |    StdDev |      Op/s |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|-------------- |---------:|----------:|----------:|----------:|-------:|-------:|------:|----------:|
| AppendCookiesBefore | 5.456 μs | 0.1469 μs | 0.4261 μs | 183,279.4 | 0.1068 | 0.0076 |     - |      4 KB |
| AppendCookiesAfter | 2.442 μs | 0.0284 μs | 0.0237 μs | 447,691.3 | 0.0801 |     - |     - |      3 KB |

#### DeleteCookies (~4x win)

|        Method |     Mean |    Error |   StdDev |     Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------- |---------:|---------:|---------:|---------:|-------:|------:|------:|----------:|
| DeleteCookiesBefore | 12.89 μs | 0.255 μs | 0.582 μs | 77,597.0 | 0.3204 |     - |     - |     11 KB |
| DeleteCookiesAfter | 3.773 μs | 0.0504 μs | 0.0421 μs | 257,194.2 | 0.1373 |     - |     - |      4 KB |

# Details

This PR took as a guidance the original comments done on ticket #31508 

**1. `DeleteCookie` calls `AppendCookie` which is designed for larger chunked cookies. However, when deleting each of the values is an empty string. It could do something much more targeted.**

It now calls directly a new public api on `ResponseCookies.cs ` who makes sure to add a collection of key pair values as cookies to the `Set-Cookier` header in an efficient manner

**2. `DeleteCookie` also could the delegate allocation using `Enumerable.Where` and use a bespoke iterator similar to `ResponseCookies.Delete`**

It now deletes the response cookies using a simple `for` loop as `ResponseCookies` class does

**3. `AppendResponseCookie` invokes `ResponseCookies.Append` which allocates a `ResponseCookieValues` instance and an intermediary `StringValues` for the header. Since we know we're running inside a loop, it could be better optimized for fewer allocations**

It now calls directly a new public api on `ResponseCookies.cs ` who makes sure to add a collection of key pair values as cookies to the `Set-Cookier` header in an efficient manner

# New API Proposal

```csharp
void Append(IDictionary<string, string> keyValuePairs, CookieOptions options);
```

As seen on `ChunkingCookieManager` the need to add several cookies at once using the same CookieOptions seems to be common, the new overload will cover this need by receiving a `IDictionary<string, string>`  whose values will be added to the `Set-Cookier` response header in an efficient manner:

- Allowing us to instante only one CookieOptions per all cookies added
- Allowing us to instante only one SetCookieHeaderValue and be reused for all Cookies to be added 
- Prevent multiple StringValues.Concat

Addresses #31508 

----
**Update 1**: Numbers were updated according to the recent changes based on the feedback

### CODE REVIEW

File: src/Http/Http/src/Internal/ResponseCookies.cs
Comment (davidfowl) [inline src/Http/Http/src/Internal/ResponseCookies.cs:104]: We want to log from these components like this.

File: src/Http/Http.Features/src/IResponseCookies.cs
Comment (davidfowl) [inline src/Http/Http.Features/src/IResponseCookies.cs:33]: This is a breaking change.

Comment (JuanShares) [reply src/Http/Http.Features/src/IResponseCookies.cs:33]: Added explanation for this on the PR and the ticket #31508, will be cool to have this but let me know what you think.

File: src/Http/Http/src/Internal/ResponseCookies.cs
Comment (JuanShares) [reply src/Http/Http/src/Internal/ResponseCookies.cs:104]: Thanks for the review, I kind of feel there is something missing after "like this", could you please tell me if I am missing something here?

File: src/Http/Http.Features/src/IResponseCookies.cs
Comment (Tratcher) [reply src/Http/Http.Features/src/IResponseCookies.cs:33]: New methods cannot be added directly to interfaces. The best mitigation in this case might be a DIM.
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/default-interface-methods

Comment (pranavkm) [reply src/Http/Http.Features/src/IResponseCookies.cs:33]: Another option is to make it an extension method that knows about `ResponseCookiesWrapper`:

```C#
public static void Append(this IResponseCookies cookies, IReadOnlyDictionary<string, string> values, CookieOptions options)
{
   if (cookies is null) { throw ArgNullException(); }
   if (cookies is ResponseCookieWrapper wrapper)
   {
      wrapper.AppendCookies(values, options);
     return;
   }

   foreach (var item in values) 
   {  
     cookies.Append(item.Key, item, value, options);
   }
}
```

It avoids changes to the interface.

File: src/Http/Http/src/Internal/ResponseCookies.cs
Comment (pranavkm) [inline src/Http/Http/src/Internal/ResponseCookies.cs:134]: Could you use a string concat (or a `StringBuilder` here? No reason to interpolate this.

File: src/Http/Http/src/Internal/ResponseCookies.cs
Comment (pranavkm) [inline src/Http/Http/src/Internal/ResponseCookies.cs:129]: I think you could borrow this from the ArrayPool. Alternatively you could simply call `Append` n times. It's effectively what the `IHeaderDictionary.Append(IList<T>)` seems to do: https://github.com/dotnet/aspnetcore/blob/5b3ba86121c6afc27ac3851308e0f00b8c809600/src/Http/Http.Extensions/src/HeaderDictionaryTypeExtensions.cs#L139-L144

File: src/Http/Http/src/Internal/ResponseCookies.cs
Comment (pranavkm) [inline src/Http/Http/src/Internal/ResponseCookies.cs:139]:
```suggestion
```

File: src/Http/Http.Features/src/IResponseCookies.cs
Comment (davidfowl) [reply src/Http/Http.Features/src/IResponseCookies.cs:33]: We need to move these interfaces to Http

File: src/Http/Http/src/Internal/ResponseCookies.cs
Comment (jkotalik) [inline src/Http/Http/src/Internal/ResponseCookies.cs:139]:
```suggestion
```

File: src/Http/Http.Features/src/IResponseCookies.cs
Comment (jkotalik) [reply src/Http/Http.Features/src/IResponseCookies.cs:33]: @pranavkm how would ResponseCookieWrapper know how to append cookies as a collection to the ResponseCookies type though?

Comment (pranavkm) [reply src/Http/Http.Features/src/IResponseCookies.cs:33]: Oh fun, didn't realize there's two different implementations. I guess default interface implementation solves that much better, doesn't it?

Comment (Tratcher) [reply src/Http/Http.Features/src/IResponseCookies.cs:33]: > We need to move these interfaces to Http

@davidfowl we can't move types without a type forward, and we can't forward in that direction without creating a circular dependency.

Why do they need to move? Because Http.Features is a netstandard 2.0 package due to the SignalR client dependency? The client wouldn't need these features so ifdefing them should be fine.

Comment (Tratcher) [inline src/Http/Http.Features/src/IResponseCookies.cs:33]: Does this need to be an `IDictionary<string, string>` or could it be less specific like IEnumerable<KeyValuePair<string, string>>`

**Comment by pranavkm** [comment]
@JuanShares looks like there's some interesting API design questions that this PR would take. Instead of randomizing you, could we discuss this as part of API review meeting and get back to you on a concrete proposal? That way we're not randomizing you too much

**Comment by JuanShares** [comment]
Cool! thank you all for the feedback, @pranavkm definitely! I will wait for the concrete proposal for the API in the meanwhile will address other feedback :D

File: src/Http/Http/src/Internal/ResponseCookies.cs
Comment (Tratcher) [reply src/Http/Http/src/Internal/ResponseCookies.cs:104]: @davidfowl this was copied from the existing code above, we can discuss it separately from this PR.
https://github.com/dotnet/aspnetcore/blob/00b551ec3ca7b50cbfe7fc0cfc91fe9e51b37b99/src/Http/Http/src/Internal/ResponseCookies.cs#L59-L71

File: src/Http/Http/src/Internal/ResponseCookies.cs
Comment (Tratcher) [reply src/Http/Http/src/Internal/ResponseCookies.cs:134]: String concat is better than StringBuilder for simple scenarios like this.

Comment (Tratcher) [reply src/Http/Http/src/Internal/ResponseCookies.cs:134]: You've skipped the escaping step from the other API:
https://github.com/dotnet/aspnetcore/blob/00b551ec3ca7b50cbfe7fc0cfc91fe9e51b37b99/src/Http/Http/src/Internal/ResponseCookies.cs#L73-L75

File: src/Http/Http.Features/src/IResponseCookies.cs
Comment (JuanShares) [reply src/Http/Http.Features/src/IResponseCookies.cs:33]: Yep, you are right `IEnumerable<KeyValuePair<string, string>>` is even more performant that `IDictionary` implementation, I am changing it

File: src/Http/Http/src/Internal/ResponseCookies.cs
Comment (JuanShares) [reply src/Http/Http/src/Internal/ResponseCookies.cs]: - I have changed to use `string.concat`
- yep, I forgot the escaping part, I have already added it back

Thanks!

File: src/Http/Http/src/Internal/ResponseCookies.cs
Comment (JuanShares) [reply src/Http/Http/src/Internal/ResponseCookies.cs]: Thanks

Comment (JuanShares) [reply src/Http/Http/src/Internal/ResponseCookies.cs]: Thanks

File: src/Http/Http/src/Internal/ResponseCookies.cs
Comment (JuanShares) [reply src/Http/Http/src/Internal/ResponseCookies.cs:105]: Correct I did not fully look at this yet and just copied from the original append, I will definitely look if this can be improved too, thanks

File: src/Http/Http/src/Internal/ResponseCookies.cs
Comment (JuanShares) [reply src/Http/Http/src/Internal/ResponseCookies.cs]: In respect to `ArrayPool`:

It sounded cool so I tried it using ` ArrayPool<string>.Shared.Rent(keyValuePairs.Count())` but it actually increases the `mean` about 100ns-200ns.

----

In respect to `Headers.Append`:
 
https://github.com/dotnet/aspnetcore/blob/9a5538162a903dc16d9de37585696a45ca4ce32d/src/Http/Http/src/Internal/ResponseCookies.cs#L133-L139

I am indeed calling it but just once at the end, I just wanted to avoid calling  `Headers.Append` multiple times because at the end it calls `StringValues.concat` and it means unnecessary allocation of `arrays`, hopefully I was able to build the full array of new cookies just like the example you gave me.
  
https://github.com/dotnet/aspnetcore/blob/c925f99cddac0df90ed0bc4a07ecda6b054a0b02/src/Http/Http.Abstractions/src/Internal/ParsingHelpers.cs#L160-L161

Thanks!

**Comment by davidfowl** [comment]
FWIW, I'm not a fan of the `IEnumerable<T>` here because it results in allocations on enumeration.

File: src/Http/Http/src/Internal/ResponseCookies.cs
Comment (gfoidl) [reply src/Http/Http/src/Internal/ResponseCookies.cs]: FWIW the C# compiler [creates string.Concat](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGAzAATakDCpA3oaQ+WRjgAykCyAFC+wNYwAngBpybUgDcAhgBsArjFG9SACxhSAJjCgA1WQoCU9RiYC8APlIASAEQ0BggL6ma0+TEc01m7XveObAG5CRyA=) for this on our behalfes. So you can use the most readable code.

File: src/Http/Http.Features/src/IResponseCookies.cs
Comment (davidfowl) [reply src/Http/Http.Features/src/IResponseCookies.cs]: > Why do they need to move? Because Http.Features is a netstandard 2.0 package due to the SignalR client dependency? The client wouldn't need these features so ifdefing them should be fine.

I want to split the IFeatureCollection from the features so we can actually change them and so we don't have http features depending on assemblies like this.

**Comment by pranavkm** [comment]
@barahonajm we think this needs some prep work on our end to allow using default interface implementations. We're going to track this as part of https://github.com/dotnet/aspnetcore/issues/31723. This PR should be unblocked once that is resolved.

@BrennanConroy / @davidfowl could we prioritize that issue to unblock this PR?

**Comment by Tratcher** [comment]
> FWIW, I'm not a fan of the `IEnumerable<T>` here because it results in allocations on enumeration.

@davidfowl IList is your recommendation to avoid this?

**Comment by davidfowl** [comment]
> @davidfowl IList is your recommendation to avoid this?

IReadOnlyList\<T\> yeah

**Comment by Tratcher** [comment]
@pranavkm I'll work with @barahonajm on this PR in parallel, we should be able to make progress using DIMs and some temporary `#if NET` statements.

**Comment by barahonajm** [comment]
@pranavkm @Tratcher thanks for the update, so it sound like we are taking the DIM path, I will start the convertion now (good challenge by the way) 

Note: github for mobile played with me, thats why you see the pr closed for a moment :D

**Comment by davidfowl** [comment]
Hmm now I want to try the ReadOnlySpan instead of IReadOnlyLisr

**Comment by barahonajm** [comment]
Well initially I went through the `IEnumerable<KeyValuePair<string, string>>` path because users could easily pass a `Dictionary` in that place, but now I believe you want a `ReadOnlySpan` as the main implementation used on `ChunkingCookierManager` and additionally add some overloads (if yes which overloads should there be?) for users to consume right?

**Comment by davidfowl** [comment]
> Well initially I went through the IEnumerable<KeyValuePair<string, string>> path because users could easily pass a Dictionary in that place, but now I believe you want a ReadOnlySpan as the main implementation used on ChunkingCookierManager and additionally add some overloads (if yes which overloads should there be?) for users to consume right?

Can you humor me and try `ReadOnlySpan<KeyValuePair<string, string>>` instead of IReadOnlyList

**Comment by barahonajm** [comment]
Yes sir, I have tried, and these are the new results (spoiler: nice suggestion!)


#### AppendCookies (~2x win, 2x less allocations)

|        Method |     Mean |     Error |    StdDev |      Op/s |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|-------------- |---------:|----------:|----------:|----------:|-------:|-------:|------:|----------:|
| AppendCookiesBefore | 5.456 μs | 0.1469 μs | 0.4261 μs | 183,279.4 | 0.1068 | 0.0076 |     - |      4 KB |
| AppendCookiesAfter[with IEnumerable] | 2.442 μs | 0.0284 μs | 0.0237 μs | 447,691.3 | 0.0801 |     - |     - |      3 KB |
| AppendCookiesAfter[with ReadOnlySpan] | 2.177 us | 0.0433 us | 0.1125 us | 2.143 us | 459,285.3 | 0.0725 | 0.0038 |     - |      2 KB |

#### DeleteCookies (~4x win, ~3x less allocations)

|        Method |     Mean |    Error |   StdDev |     Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------- |---------:|---------:|---------:|---------:|-------:|------:|------:|----------:|
| DeleteCookiesBefore | 12.89 μs | 0.255 μs | 0.582 μs | 77,597.0 | 0.3204 |     - |     - |     11 KB |
| DeleteCookiesAfter[with IEnumerable] | 3.773 μs | 0.0504 μs | 0.0421 μs | 257,194.2 | 0.1373 |     - |     - |      4 KB |
| DeleteCookies[with ReadOnlySpan] | 3.155 us | 0.0625 us | 0.0973 us | 3.144 us | 316,973.5 | 0.1183 |      - |     - |      4 KB |

**Comment by davidfowl** [comment]
This is the new API 😄

**Comment by barahonajm** [comment]
@Tratcher Sorry for the delay, I finally was able to get back to this, I was thinking on the DIM implementation thought this these days and come to the conclusion it should be as simple as calling the normal `Append` are you agree with this? (This is my first use of DIM :D)

https://github.com/dotnet/aspnetcore/blob/df9ee2fa8ef44dac544ebafc0ee7079c328f0267/src/Http/Http.Features/src/IResponseCookies.cs#L31-L44

FYI @davidfowl @pranavkm

File: src/Http/Http/src/Internal/ResponseCookies.cs
Comment (Tratcher) [inline src/Http/Http/src/Internal/ResponseCookies.cs:134]: Style: long line, pull out `_enableCookieNameEncoding ? Uri.EscapeDataString(keyValuePair.Key) : keyValuePair.Key` to a local `var key` on its own line.

File: src/Security/CookiePolicy/src/ResponseCookiesWrapper.cs
Comment (Tratcher) [inline src/Security/CookiePolicy/src/ResponseCookiesWrapper.cs:164]:
```suggestion
            var nonSuppressedValues = new List<KeyValuePair<string, string>>(keyValuePairs.Length);
```

File: src/Security/CookiePolicy/test/CookieChunkingTests.cs
Comment (Tratcher) [inline src/Security/CookiePolicy/test/CookieChunkingTests.cs:160]: The Cookie header is a single line and doesn't include any of the attributes.
```suggestion
                "TestCookie=chunks-7, TestCookieC1=abcdefghi, ..."
```

File: src/Security/CookiePolicy/test/CookieChunkingTests.cs
Comment (Tratcher) [inline src/Security/CookiePolicy/test/CookieChunkingTests.cs:190]: Verify the new values.

File: src/Security/perf/Microbenchmarks/ChunkingCookieManagerBenchmark.cs
Comment (Tratcher) [inline src/Security/perf/Microbenchmarks/ChunkingCookieManagerBenchmark.cs:35]: Wrong format, see my comments in the tests.

**Review by Tratcher — COMMENTED** [review]
The functional part of this looks fine, I've added some minor comments. I'll follow up about the PublicAPI cross targeting issues to see if this is the expected pattern. Worst case we get someone to do #31723 first and it becomes moot.

File: src/Http/Http.Features/src/Microsoft.AspNetCore.Http.Features.csproj
Comment (dougbu) [inline src/Http/Http.Features/src/Microsoft.AspNetCore.Http.Features.csproj:13]: Agree w/ the newline removal but not the excess indentation
```suggestion
 <ItemGroup>
```

File: src/Security/perf/Microbenchmarks/Microsoft.AspNetCore.Security.Microbenchmarks.csproj
Comment (dougbu) [inline src/Security/perf/Microbenchmarks/Microsoft.AspNetCore.Security.Microbenchmarks.csproj:17]: Nit: We tend to sort `@(Reference)` items which would place this second, after the already-out-of-place `BenchmarkDotNet` reference. Up to you whether to include this in your next iteration.

**Review by dougbu — APPROVED** [review]
From a build perspective but please address https://github.com/dotnet/aspnetcore/pull/31625/files?file-filters%5B%5D=.csproj&file-filters%5B%5D=.txt#r616226305 and get additional approvals before squish / merge.

File: src/Http/Http/src/Internal/ResponseCookies.cs
Comment (barahonajm) [reply src/Http/Http/src/Internal/ResponseCookies.cs]: Fixed

File: src/Security/CookiePolicy/test/CookieChunkingTests.cs
Comment (barahonajm) [reply src/Security/CookiePolicy/test/CookieChunkingTests.cs]: I changed it to not include the attributes, but I still have have it in multiple lines as the other tests for readability

File: src/Security/perf/Microbenchmarks/Microsoft.AspNetCore.Security.Microbenchmarks.csproj
Comment (barahonajm) [reply src/Security/perf/Microbenchmarks/Microsoft.AspNetCore.Security.Microbenchmarks.csproj]: Thanks, it was resolved

File: src/Http/Http.Features/src/Microsoft.AspNetCore.Http.Features.csproj
Comment (barahonajm) [inline src/Http/Http.Features/src/Microsoft.AspNetCore.Http.Features.csproj:13]:
```suggestion
  <ItemGroup>
```

File: src/Http/Http.Features/src/Microsoft.AspNetCore.Http.Features.csproj
Comment (barahonajm) [reply src/Http/Http.Features/src/Microsoft.AspNetCore.Http.Features.csproj]: This was the trickiest part to resolve, for some reason my computer showed 2 spaces but on GitHub it was clearly 3, thanks

File: src/Security/CookiePolicy/test/CookieChunkingTests.cs
Comment (barahonajm) [reply src/Security/CookiePolicy/test/CookieChunkingTests.cs:190]: Thanks, done

File: src/Http/Http.Features/src/IResponseCookies.cs
Comment (Tratcher) [inline src/Http/Http.Features/src/IResponseCookies.cs:33]:
```suggestion
        /// Add elements of specified collection as cookies.
```

**Review by Tratcher — APPROVED** [review]

**Comment by Tratcher** [comment]
Thanks, that had a lot of build complications to deal with.

**Pull request merged**
